File: app/src/main/java/crux/bphc/cms/fragments/DiscussionFragment.java
package crux.bphc.cms.fragments;


import android.os.Bundle;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.Arrays;

import crux.bphc.cms.R;
import crux.bphc.cms.app.MyApplication;
import crux.bphc.cms.helper.FileManager;
import crux.bphc.cms.helper.HtmlTextView;
import crux.bphc.cms.helper.PropertiesAlertDialog;
import crux.bphc.cms.models.forum.Attachment;
import crux.bphc.cms.models.forum.Discussion;
import io.realm.Realm;


public class DiscussionFragment extends Fragment {

    private  String mCourseName;
    private int id;

    private Realm realm;
    private FileManager mFileManager;

    private ImageView mUserPic;
    private TextView mSubject;
    private TextView mUserName;
    private TextView mTimeModified;
    private HtmlTextView mMessage;
    private LinearLayout mAttachmentContainer;

    private MoreOptionsFragment.OptionsViewModel moreOptionsViewModel;

    public DiscussionFragment() {
        // Required empty public constructor
    }

    public static DiscussionFragment newInstance(int id, String mCourseName) {
        DiscussionFragment fragment = new DiscussionFragment();
        Bundle args = new Bundle();
        args.putInt("id", id);
        args.putString("courseName", mCourseName);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            id = getArguments().getInt("id");
            mCourseName = getArguments().getString("courseName");
        }
        realm = MyApplication.getInstance().getRealmInstance();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_discussion, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        moreOptionsViewModel = new ViewModelProvider(requireActivity()).get(MoreOptionsFragment.OptionsViewModel.class);

        Discussion discussion = realm.where(Discussion.class).equalTo("id", id).findFirst();

        mFileManager = new FileManager(getActivity(), mCourseName);
        mFileManager.registerDownloadReceiver();
        mFileManager.setCallback(filename -> {
            int child = mAttachmentContainer.getChildCount();
            for (int i = 0; i < child; i++) {
                View childView = mAttachmentContainer.getChildAt(i);
                TextView fileNameTextView = childView.findViewById(R.id.fileName);
                if (fileNameTextView != null &&
                        fileNameTextView.getText().toString().equalsIgnoreCase(filename)) {
                    ImageView downloadIcon = childView.findViewById(R.id.downloadButton);
                    downloadIcon.setImageResource(R.drawable.eye);
                    ImageView ellipsis = childView.findViewById(R.id.more);
                    ellipsis.setVisibility(View.VISIBLE);
                    break;
                }
            }

            Attachment attachment = discussion.getAttachments().where().equalTo("filename", filename).findFirst();
            if (attachment != null)
                mFileManager.openDiscussionAttachment(attachment);
        });

        mUserPic = view.findViewById(R.id.user_pic);
        Picasso.get().load(discussion.getUserpictureurl()).into(mUserPic);

        mSubject = view.findViewById(R.id.subject);
        mSubject.setText(discussion.getSubject());

        mUserName = view.findViewById(R.id.user_name);
        mUserName.setText(discussion.getUserfullname());

        mTimeModified = view.findViewById(R.id.modified_time);
        mTimeModified.setText(ForumFragment.formatDate(discussion.getTimemodified()));

        mMessage = view.findViewById(R.id.message);
        mMessage.setText(HtmlTextView.parseHtml(discussion.getMessage()));
        mMessage.setMovementMethod(LinkMovementMethod.getInstance());

        mAttachmentContainer = view.findViewById(R.id.attachments);
        LayoutInflater inflater = LayoutInflater.from(getContext());

        if (discussion.getAttachments().size() == 0) mAttachmentContainer.setVisibility(View.GONE);

        for (final Attachment attachment : discussion.getAttachments()) {
            View attachmentView = inflater.inflate(
                    R.layout.row_attachment_detail_forum,
                    mAttachmentContainer);

            TextView fileName = attachmentView.findViewById(R.id.fileName);
            fileName.setText(attachment.getFilename());

            LinearLayout clickWrapper = attachmentView.findViewById(R.id.clickWrapper);
            ImageView download = attachmentView.findViewById(R.id.downloadButton);
            ImageView ellipsis = attachmentView.findViewById(R.id.more);

            boolean downloaded = mFileManager.isDiscussionAttachmentDownloaded(attachment);
            if (downloaded) {
                download.setImageResource(R.drawable.eye);
                ellipsis.setVisibility(View.VISIBLE);
            } else {
                download.setImageResource(R.drawable.download);
                ellipsis.setVisibility(View.GONE);
            }

            clickWrapper.setOnClickListener(v -> {
                if (!downloaded) {
                    Toast.makeText(getActivity(), "Downloading file - " + attachment.getFilename(), Toast.LENGTH_SHORT).show();
                    mFileManager.downloadDiscussionAttachment(attachment, discussion.getSubject(), mCourseName);
                } else {
                    mFileManager.openDiscussionAttachment(attachment);
                }
            });

            ellipsis.setOnClickListener(v -> {
                // Check if downloaded once again, for consistency (user downloaded and then opens ellipsis immediately)
                boolean isDownloaded = mFileManager.isDiscussionAttachmentDownloaded(attachment);
                if (isDownloaded) {
                    ArrayList<MoreOptionsFragment.Option> options = new ArrayList<>();
                    Observer<MoreOptionsFragment.Option> observer;  // to handle the selection

                    options.addAll(Arrays.asList(
                            new MoreOptionsFragment.Option(0, "View", R.drawable.eye),
                            new MoreOptionsFragment.Option(1, "Re-Download", R.drawable.download),
                            new MoreOptionsFragment.Option(2, "Share", R.drawable.ic_share),
                            new MoreOptionsFragment.Option(3, "Properties", R.drawable.ic_info)
                    ));

                    observer = option -> {
                        if (option == null) return;
                        switch (option.getId()) {
                            case 0:
                                mFileManager.openDiscussionAttachment(attachment);
                                break;
                            case 1:
                                Toast.makeText(getActivity(), "Downloading file - " + attachment.getFilename(),
                                        Toast.LENGTH_SHORT).show();
                                mFileManager.downloadDiscussionAttachment(attachment, discussion.getSubject(),
                                        mCourseName);
                                break;
                            case 2:
                                mFileManager.shareDiscussionAttachment(attachment);
                                break;
                            case 3:
                                new PropertiesAlertDialog(getContext(), attachment).show();
                        }
                        moreOptionsViewModel.getSelection().removeObservers((AppCompatActivity) getContext());
                        moreOptionsViewModel.clearSelection();
                    };

                    MoreOptionsFragment fragment = MoreOptionsFragment.newInstance(attachment.getFilename(), options);
                    fragment.show(((AppCompatActivity) getContext()).getSupportFragmentManager(),
                            fragment.getTag());
                    moreOptionsViewModel.getSelection().observe((AppCompatActivity) getContext(), observer);
                }
            });
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mFileManager.unregisterDownloadReceiver();
    }
}

File: app/src/main/java/crux/bphc/cms/fragments/DiscussionFragment.kt
package crux.bphc.cms.fragments

import android.os.Bundle
import android.text.method.LinkMovementMethod
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.annotation.MainThread
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.bumptech.glide.Glide
import crux.bphc.cms.R
import crux.bphc.cms.app.Urls
import crux.bphc.cms.core.FileManager
import crux.bphc.cms.databinding.FragmentDiscussionBinding
import crux.bphc.cms.databinding.RowAttachmentDetailForumBinding
import crux.bphc.cms.fragments.MoreOptionsFragment.Companion.newInstance
import crux.bphc.cms.fragments.MoreOptionsFragment.OptionsViewModel
import crux.bphc.cms.helper.CourseDataHandler
import crux.bphc.cms.helper.CourseRequestHandler
import crux.bphc.cms.models.forum.Attachment
import crux.bphc.cms.models.forum.Discussion
import crux.bphc.cms.utils.Utils
import crux.bphc.cms.widgets.PropertiesAlertDialog
import io.realm.Realm
import io.realm.RealmList
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.IOException
import java.util.*

class DiscussionFragment : Fragment() {

    private lateinit var realm: Realm
    private lateinit var binding: FragmentDiscussionBinding
    private lateinit var fileManager: FileManager
    private lateinit var moreOptionsViewModel: OptionsViewModel
    private lateinit var discussion: Discussion

    private var courseId: Int = 0
    private var forumId: Int = 0
    private var discussionId: Int = 0
    private var mCourseName: String = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        courseId = requireArguments().getInt(COURSE_ID_KEY, -1)
        forumId = requireArguments().getInt(FORUM_ID_KEY, -1)
        discussionId = requireArguments().getInt(DISCUSSION_ID_KEY, -1)
        mCourseName = requireArguments().getString(COURSE_NAME_KEY, "")

        fileManager = FileManager(requireActivity(), mCourseName) { oneFileDownloaded(it) }
        realm = Realm.getDefaultInstance()
        binding = FragmentDiscussionBinding.inflate(layoutInflater)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        moreOptionsViewModel = ViewModelProvider(requireActivity())[OptionsViewModel::class.java]

        binding.refresh.isEnabled = false // Disable swiping, for now
        binding.message.movementMethod = LinkMovementMethod.getInstance()

        fileManager.registerDownloadReceiver()

        val discussion = realm.where(Discussion::class.java).equalTo("discussionId", discussionId)
            .findFirst()
        if (discussion != null) {
            setDiscussion(discussion)
        } else {
            refreshContent(forumId, discussionId)
        }
    }

    @MainThread
    private fun setDiscussion(discussion: Discussion) {
        this.discussion = discussion
        binding.content.visibility = View.VISIBLE
        binding.empty.visibility = View.GONE
        binding.refresh.isRefreshing = false

        binding.subject.text = discussion.subject
        binding.userName.text = discussion.userFullName
        binding.modifiedTime.text = Utils.formatDate(discussion.timeModified)
        binding.message.text = discussion.message
        Glide.with(requireContext())
            .load(Urls.getProfilePicUrl(discussion.userPictureUrl))
            .into(binding.userPic)

        val attachments = discussion.attachments
        if (attachments.isNotEmpty()) {
            val inflater = LayoutInflater.from(requireContext())
            binding.attachments.visibility = View.VISIBLE

            for (attachment in discussion.attachments) {
                val attachmentBinding = RowAttachmentDetailForumBinding.inflate(inflater, binding.attachments, true)
                attachmentBinding.let {
                    it.name.text = attachment.fileName
                    if (fileManager.isDiscussionAttachmentDownloaded(attachment)) {
                        it.download.setImageResource(R.drawable.eye)
                        it.more.visibility = View.VISIBLE
                    } else {
                        it.download.setImageResource(R.drawable.outline_file_download_24)
                        it.more.visibility = View.GONE
                    }
                    it.clickWrapper.setOnClickListener {
                        if (!fileManager.isDiscussionAttachmentDownloaded(attachment)) {
                            downloadAttachment(attachment)
                        } else {
                            fileManager.openDiscussionAttachment(attachment)
                        }
                    }
                    it.more.setOnClickListener {
                        if (fileManager.isDiscussionAttachmentDownloaded(attachment)) {
                            val options = ArrayList(listOf(
                                MoreOptionsFragment.Option(0, "View", R.drawable.eye),
                                MoreOptionsFragment.Option(1, "Re-Download", R.drawable.outline_file_download_24),
                                MoreOptionsFragment.Option(2, "Share", R.drawable.ic_share),
                                MoreOptionsFragment.Option(3, "Properties", R.drawable.ic_info)
                            ))
                            val observer: Observer<MoreOptionsFragment.Option?> = Observer { option: MoreOptionsFragment.Option? ->
                                option ?: return@Observer
                                when (option.id) {
                                    0 -> fileManager.openDiscussionAttachment(attachment)
                                    1 -> {
                                        downloadAttachment(attachment)
                                    }
                                    2 -> fileManager.shareDiscussionAttachment(attachment)
                                    3 -> PropertiesAlertDialog(requireActivity(), attachment).show()
                                }
                                moreOptionsViewModel.selection.removeObservers(requireActivity())
                                moreOptionsViewModel.clearSelection()
                            }
                            val fragment = newInstance(attachment.fileName, options)
                            fragment.show(requireActivity().supportFragmentManager, fragment.tag)
                            moreOptionsViewModel.selection.observe(requireActivity(), observer)
                        }
                    }
                    if(attachment != discussion.attachments.last()) {
                        it.barrier.setDpMargin(4)
                    }
                }
            }
        } else {
            binding.attachments.visibility = View.GONE
        }
    }

    private fun refreshContent(forumId: Int, discussionId: Int) {
        binding.refresh.isRefreshing = true
        CoroutineScope(Dispatchers.IO).launch {
            val realm = Realm.getDefaultInstance()
            val courseDataHandler = CourseDataHandler(realm)
            val courseRequestHandler = CourseRequestHandler()

            try {
                val discussions = courseRequestHandler.getForumDicussionsSync(forumId)
                discussions.forEach { it.forumId = forumId }
                courseDataHandler.setForumDiscussions(forumId, discussions)

                val discussion = discussions.firstOrNull { it.discussionId == discussionId }
                CoroutineScope(Dispatchers.Main).launch {
                    if (discussion != null) {
                        binding.empty.visibility = View.GONE
                        binding.refresh.isRefreshing = false
                        setDiscussion(discussion)
                    } else {
                        binding.empty.visibility = View.VISIBLE
                        binding.refresh.isRefreshing = false
                        Toast.makeText(requireContext(), getString(R.string.net_req_failed),
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }
            } catch (e: IOException) {
                CoroutineScope(Dispatchers.Main).launch {
                    binding.empty.visibility = View.VISIBLE
                    binding.refresh.isRefreshing = false
                    Toast.makeText(requireContext(), getString(R.string.net_req_failed),
                        Toast.LENGTH_SHORT
                    ).show()
                }

            }
            realm.close()
        }
    }

    private fun downloadAttachment(attachment: Attachment) {
        Toast.makeText(activity, getString(R.string.downloading_file) + attachment.fileName,
            Toast.LENGTH_SHORT).show()
        fileManager.downloadDiscussionAttachment(attachment, discussion.subject)
    }

    private fun oneFileDownloaded(filename: String) {
        val child = binding.attachments.childCount
        // Count starts from 2 as the layout contains two extra views before further rows are attached
        for (i in 2 until child) {
            val childView = binding.attachments.getChildAt(i) ?: break
            val attachmentBinding = RowAttachmentDetailForumBinding.bind(childView)

            if (attachmentBinding.name.text.toString().equals(filename, true)) {
                attachmentBinding.download.setImageResource(R.drawable.eye)
                attachmentBinding.more.visibility = View.VISIBLE
            }
        }

        val attachments = discussion.attachments
        val attachment = attachments.firstOrNull { it.fileName == filename }
        attachment?.let { fileManager.openDiscussionAttachment(it) }

    }

    override fun onDestroyView() {
        super.onDestroyView()
        fileManager.unregisterDownloadReceiver()
    }

    override fun onDestroy() {
        super.onDestroy()
        realm.close()
    }

    companion object {

        private const val COURSE_ID_KEY = "courseId"
        private const val COURSE_NAME_KEY = "contextUrl"
        private const val FORUM_ID_KEY = "forumId"
        private const val DISCUSSION_ID_KEY = "discussionId"

        @JvmStatic
        fun newInstance(
            courseId: Int,
            forumId: Int,
            discussionId: Int,
            mCourseName: String,
        ): DiscussionFragment {
            val fragment = DiscussionFragment()
            val args = Bundle()
            args.putInt(COURSE_ID_KEY, courseId)
            args.putInt(FORUM_ID_KEY, forumId)
            args.putInt(DISCUSSION_ID_KEY, discussionId)
            args.putString(COURSE_NAME_KEY, mCourseName)
            fragment.arguments = args
            return fragment
        }
    }
}